### ISSUE TITLE
Invalid Parameter Type: Expected params. Payload to be a string, Buffer, Stream, Blob, or typed array object.

### ISSUE THREAD

**Issue by nephyo** [alexcasalboni/aws-lambda-power-tuning#97]
Hi,
I have an issue when providing this type of weighted payload value:
```json
{
  "lambdaARN": "my_lambda_ARN",
  "powerValues": [
    128,
    256,
    512,
    1024,
    2048,
    3008
  ],
  "num": 10,
  "payload": [
    {
      "payload": {
        "body-json": {
          "dateRanges": [
            {
              "start": "2020-05-01",
              "end": "2020-05-31"
            },
            {
              "start": "2020-04-01",
              "end": "2020-04-30"
            }
          ]
        },
        "params": {
          "path": {
            "tenantId": "DA29004"
          },
          "querystring": {},
          "header": {
            "Authorization": "Bearer some_bearer"
          }
        }
      },
      "weight": 1
    },
    {
      "payload": {
        "body-json": {
          "dateRanges": [
            {
              "start": "2020-05-01",
              "end": "2020-05-31"
            },
            {
              "start": "2020-03-01",
              "end": "2020-03-31"
            }
          ]
        },
        "params": {
          "path": {
            "tenantId": "DA29004"
          },
          "querystring": {},
          "header": {
            "Authorization": "Bearer some_bearer"
          }
        }
      },
      "weight": 1
    },
    {
      "payload": {
        "body-json": {
          "dateRanges": [
            {
              "start": "2020-05-01",
              "end": "2020-05-31"
            },
            {
              "start": "2020-02-01",
              "end": "2020-02-29"
            }
          ]
        },
        "params": {
          "path": {
            "tenantId": "DA29004"
          },
          "querystring": {},
          "header": {
            "Authorization": "Bearer some_bearer"
          }
        }
      },
      "weight": 1
    }
  ],
  "parallelInvocation": false,
  "strategy": "balanced"
}
```

It results with:
```json
"error": {
    "Error": "InvalidParameterType",
    "Cause": "{\"errorType\":\"InvalidParameterType\",\"errorMessage\":\"Expected params.Payload to be a string, Buffer, Stream, Blob, or typed array object\",\"trace\":[\"InvalidParameterType: Expected params.Payload to be a string, Buffer, Stream, Blob, or typed array object\",\"    at ParamValidator.fail (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:50:37)\",\"    at ParamValidator.validatePayload (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:267:10)\",\"    at ParamValidator.validateScalar (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:133:21)\",\"    at ParamValidator.validateMember (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:94:21)\",\"    at ParamValidator.validateStructure (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:75:14)\",\"    at ParamValidator.validateMember (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:88:21)\",\"    at ParamValidator.validate (/var/runtime/node_modules/aws-sdk/lib/param_validator.js:34:10)\",\"    at Request.VALIDATE_PARAMETERS (/var/runtime/node_modules/aws-sdk/lib/event_listeners.js:126:42)\",\"    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\",\"    at callNextListener (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:96:12)\"]}"
  }
```

Now, I do not know why this is happening and what it actually means. There are three objects in payload array. When I remove one. Execution is successful. It is payload for Lambda used for POST requests.

**Comment by alexcasalboni** [comment]
Hi @nephyo, thanks for reporting this.

I've just tested your input against the test suit and in a production environment as well - unfortunately I could not reproduce the error.

It's quite weird because the `params.Payload` value is always converted into string no matter what, so I can't think of a root cause without diving deeper with your help.

A few questions so we can identify the issue together:

1. When did you deploy the application/stack? Did you deploy via SAR or via SAM CLI?
2. Could you share a screenshot of the state machine execution diagram? (Which step is failing?)
3. Is this stack trace generated by the `Executor` function or is it an error generated by your own function?
4. Does this happen 100% of the times when you use three objects? (Does it look like a race condition or is it a stable behaviour?)

Thanks :)

**Comment by nephyo** [comment]
Hi @alexcasalboni, thanks for reply.

1. It was deployed on Jun 9th 2020 via SAR.
2. ![image](https://user-images.githubusercontent.com/7772439/84662032-847d3f80-af1b-11ea-98f4-0b948c90a9c9.png)
3. Generated by Executor function.
4. Yes, It happens every time when I use three objects.

Thank you.

**Comment by alexcasalboni** [comment]
I see, thanks for the quick response.

Because I've just published a brand new version [on SAR](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:451282441545:applications~aws-lambda-power-tuning) (v3.3.0) on [Jun 10th](https://twitter.com/alex_casalboni/status/1270765001775689728), the first thing I'd try is deleting the stack and deploying the latest version. This way, we know whether the problem can be reproduced in the current code base or not.

**Comment by nephyo** [comment]
I have deployed the latest version of the tool via SAR. I can confirm that the issue is in this version as well.

**Comment by alexcasalboni** [comment]
Ok, good news. I managed to reproduce it by using the whole input (the one you shared above).

The bug seems to be related to the very low value of `num` (in your case, `10`).

This is definitely a bug that needs to be fixed. For now, you should be able to work around it simply by using a higher value for `num`, for example `30`.

I'll run some more tests later and keep you posted here.

**Comment by alexcasalboni** [comment]
Update: I've identified the underlying issue and I'm now in the process of fixing the bug and adding more unit tests to avoid future regressions.

**Comment by nephyo** [comment]
Awesome. Thank you. I can confirm that when I increase the `num` value, execution is successful.

**Comment by alexcasalboni** [comment]
@nephyo I'm glad the workaround helped.

It turned out the bug wasn't related to the low value of num per se, as the bug appeared even with higher values. In fact, it was linked to the relationship between `num` and the number of weighted payloads. Just a simple algorithmic mistake in the function that generates and allocates payloads.

I hope this bug didn't cause you too much trouble.

I'm going to merge #98 and publish a new SAR app later today.

**Comment by nephyo** [comment]
Alright. I'm glad that you were able to identify it so quickly.
Thanks again.

**Comment by alexcasalboni** [comment]
The new version `3.3.1` is now available on SAR :)

**Comment by nephyo** [comment]
Unfortunately, for me, the issue is still there. I got 26 payload objects, in array. I'm still seeing the same error message. `Num` value is set to 30.

**Comment by alexcasalboni** [comment]
Sorry @nephyo, you are absolutely right.

I've just opened a new PR to cover the missing cases. Merging soon.

This time it should be for good :)

**Comment by alexcasalboni** [comment]
Done. And published to SAR as version `3.3.2`.

**Comment by nephyo** [comment]
Thank you. It works now.
